### ISSUE TITLE
Gemma 4: PagedAttention paged decode kernel ignores sliding window constraint (correctness bug)

### ISSUE THREAD

**Issue by rnett** [EricLBuehler/mistral.rs#2094]
## Summary

When running Gemma 4 with PagedAttention enabled, the paged attention decode kernel is called **without the sliding window size** for the 35/42 sliding window attention layers. This means the kernel attends over the full context instead of the 512-token window, producing incorrect outputs for any generation beyond 512 tokens.

## Environment

| Field | Value |
|-------|-------|
| **GPU** | NVIDIA GeForce RTX 5080 |
| **Compute Capability** | 12.0 |
| **VRAM** | 16303 MiB |
| **GPU Driver** | 595.97 |
| **OS** | Ubuntu 24.04.4 LTS (Noble Numbat), WSL2 on Windows 11 |
| **CUDA** | 13.2 (nvcc V13.2.51, built 2026-03-02) |
| **cuDNN** | 9.20.0.48 (`cudnn9-cuda-13-2`) |
| **Rust** | 1.94.1 (e408947bf 2026-03-25), LLVM 21.1.8 |
| **Install method** | `cargo install mistralrs-cli --version 0.8.1 --features cuda,cudnn,flash-attn` (compiled from crates.io during install) |
| **mistralrs version** | 0.8.1 |
| **Cargo features** | `cuda`, `cudnn`, `flash-attn` |
| **Model** | google/gemma-4-E4B-it |
| **Affected layers** | 35 of 42 (all `sliding_attention` type layers) |
| **Sliding window size** | 512 tokens |

## Gemma 4 E4B Architecture (from model config)

```
layer_types: ["sliding_attention" × 5, "full_attention", ...] (pattern repeats 7x)
sliding_window: 512
global_head_dim: 512
head_dim: 256 (sliding layers)
sliding_window_pattern: 6
```

35 out of 42 layers are `sliding_attention`; 7 are `full_attention` (every 6th).

### Where the mask is suppressed (correct behaviour)

In `mistralrs-core/src/vision_models/gemma4/text.rs`, around line 1542:

```rust
let sliding_attention_mask = if metadata
    .as_ref()
    .map(|(_, meta)| meta.is_first_prompt_chunk)
    .unwrap_or(true)
{
    sliding_attention_mask      // prefill: pass the Custom mask
} else {
    AttentionMask::None         // decode: None — correct, lets paged kernel handle it
};
```

During decode `is_first_prompt_chunk = false`, so `sliding_attention_mask = AttentionMask::None`. This is the right intent — the paged kernel is supposed to handle the windowing.

### Where the window is not passed to the kernel

In `mistralrs-core/src/paged_attention/layers/paged_attention.rs`, around line 402, the decode path calls:

```rust
let res = paged_attention(
    &query,
    self.k_scale.as_ref(),
    self.v_scale.as_ref(),
    key_cache.as_ref().unwrap(),
    value_cache.as_ref().unwrap(),
    resolve_block_tables(&dev).unwrap(),
    resolve_context_lens(&dev).unwrap(),
    alibi_slopes.as_ref(),
    if use_full {
        input_metadata.full_max_context_len.unwrap()
    } else {
        input_metadata.max_context_len.unwrap()   // ← full context len, not windowed
    },
    sdpa_params.softmax_scale,
    sdpa_params.softcap.unwrap_or(1.0f32),
    sdpa_params.sinks.as_ref(),
)?;
```

The `paged_attention` kernel has **no `sliding_window` parameter**. The `max_context_len` passed for sliding window layers is the full sequence length, not capped to 512.

The block selection logic at lines 175–180:

```rust
// For models with per-layer sliding windows (GPT-OSS, Gemma2):
// - Full-attention layers (sliding_window == None) use the full block tables.
// - Sliding-window layers (sliding_window == Some) use the windowed block tables.
let use_full =
    sdpa_params.sliding_window.is_none() && input_metadata.full_block_tables.is_some();
```

This correctly selects windowed vs full block tables. However, even if `block_tables` only contains blocks within the window, `max_context_len` is still the full sequence length, which may cause the kernel to read beyond the allocated windowed blocks.

## Impact

- **Correctness:** For sequences longer than 512 tokens, sliding window attention layers attend to tokens outside the window, breaking the model's expected attention pattern. This can silently degrade output quality without any error or warning.
- **Affected models:** Any model with per-layer sliding window attention used with PagedAttention — confirmed for Gemma 4 E4B/26B, likely also affects other Gemma 4 variants and any other model using `sliding_window_pattern`.

## Suggested Fix

The `paged_attention` CUDA kernel should accept a `sliding_window: Option<usize>` parameter and enforce the window during the block-table scan. Alternatively, the scheduler should ensure `context_lens` for sliding window layers are capped to `min(actual_pos, sliding_window)` before passing to the kernel.

**Comment by EricLBuehler** [comment]
Hey @rnett! Thanks for the issue. However, this is actually not a bug and we handle sliding window in this case properly at the input-metadata level:

Decode path windowing (mistralrs-core/src/pipeline/inputs_processor.rs:618-666):
- For sliding-window layers, the scheduler emits windowed block_tables (truncated to start at seq.len().saturating_sub(sliding_window) / block_size) and capped paged_attn_context_lens (seq.len() -
block_aligned_start).
- For full-attention layers, it emits separate full_block_tables / full_context_lens covering the entire context.

See:
>https://github.com/EricLBuehler/mistral.rs/blob/3b3539173309473f04c1caa177610139764b8f84/mistralrs-core/src/paged_attention/layers/paged_attention.rs#L179-L195
